near/omni-bridge/src/lib.rs

```diff
@@ -883,6 +894,149 @@ impl Contract {
+    ) -> PromiseOrPromiseIndexOrValue<U128> {
+        // Verify this is indeed a UTXO chain token
+        let origin_chain = self
+            .get_utxo_chain_by_token(&token_id)
+            .sdk_expect("ERR_TOKEN_NOT_FROM_UTXO_CHAIN");
+
+        env::log_str(
+            &OmniBridgeEvent::UtxoTransferEvent {
+                transfer_message: utxo_fin_transfer_msg.clone(),
+            }
+            .to_log_string(),
+        );
+
+        let fast_transfer = FastTransfer::from_utxo_transfer(
+            utxo_fin_transfer_msg.clone(),
+            token_id.clone(),
+            amount,
+            origin_chain,
+        );
+
+        if let Some(status) = self.get_fast_transfer_status(&fast_transfer.id()) {
+            return self.utxo_fin_transfer_fast(fast_transfer, status);
+        }
+
+        if let OmniAddress::Near(recipient) = utxo_fin_transfer_msg.recipient {
+            self.utxo_fin_transfer_to_near(recipient, token_id, amount, &utxo_fin_transfer_msg.msg)
+                .into()
+        } else {
+            self.utxo_fin_transfer_to_other_chain(
+                token_id,
+                amount,
+                utxo_fin_transfer_msg,
+                signer_id,
+            )
+        }
+    }
+
+    fn utxo_fin_transfer_fast(
+        &mut self,
+        fast_transfer: FastTransfer,
+        fast_transfer_status: FastTransferStatus,
+    ) -> PromiseOrPromiseIndexOrValue<U128> {
+        require!(
+            !fast_transfer_status.finalised,
+            "ERR_FAST_TRANSFER_ALREADY_FINALISED"
+        );
+
+        let amount = if fast_transfer.recipient.get_chain() == ChainKind::Near {
+            self.remove_fast_transfer(&fast_transfer.id());
+            fast_transfer.amount
+        } else {
+            require!(
+                !fast_transfer_status.finalised,
+                "ERR_FAST_TRANSFER_ALREADY_FINALISED"
+            );
+            self.mark_fast_transfer_as_finalised(&fast_transfer.id());
+            // With transfers to other chain the fee will be claimed after finalization on the destination chain
+            U128(fast_transfer.amount.0 - fast_transfer.fee.fee.0)
+        };
+
+        self.send_tokens(
+            fast_transfer.token_id,
+            fast_transfer_status.relayer,
+            amount,
+            "",
+        );
+
+        PromiseOrPromiseIndexOrValue::Value(U128(0))
+    }
+
+    fn utxo_fin_transfer_to_near(
+        &self,
+        recipient: AccountId,
+        token_id: AccountId,
+        amount: U128,
+        msg: &str,
+    ) -> Promise {
+        let deposit_action = StorageDepositAction {
+            account_id: recipient.clone(),
+            token_id: token_id.clone(),
+            storage_deposit_amount: None,
+        };
+
+        // We send the recipient full amount including fee, because fee is only taken in case of fast transfers
+        Self::check_or_pay_ft_storage(&deposit_action, &mut NearToken::from_yoctonear(0)).then(
```

Comment:
Why do we even need this check? Here we’re just calling storage_balance_of, and we don’t even check the result anywhere.

Reply:
Because we can't fail which can happen if recipient doesn't have storage deposited for utxo token. Will fix the check